be/src/exprs/vectorized/array_functions.cpp

```diff
@@ -884,6 +885,148 @@ ColumnPtr ArrayFunctions::array_arithmetic(const Columns& columns) {
+        } else {
+            hash_set->emplace(item.get<CppType>());
+        }
+    }
+
+    auto& dest_data_column = dest_column->elements_column();
+    auto& dest_offsets = dest_column->offsets_column()->get_data();
+
+    if (has_null) {
+        dest_data_column->append_nulls(1);
+    }
+
+    auto iter = hash_set->begin();
+    while (iter != hash_set->end()) {
+        dest_data_column->append_datum(*iter);
+        iter++;
+    }
+    dest_offsets.emplace_back(dest_offsets.back() + hash_set->size() + has_null);
+}
+
+template <PrimitiveType PT, typename HashSet>
+ColumnPtr ArrayFunctions::array_distinct(const Columns& columns) {
+    DCHECK_EQ(columns.size(), 1);
+
+    size_t chunk_size = columns[0]->size();
+    ColumnPtr src_column = ColumnHelper::unpack_and_duplicate_const_column(chunk_size, columns[0]);
+    ColumnPtr dest_column = src_column->clone_empty();
+
+    HashSet hash_set;
+
+    if (columns[0]->is_nullable()) {
+        const auto* src_nullable_column = down_cast<const NullableColumn*>(src_column.get());
+        const auto* src_data_column = down_cast<const ArrayColumn*>(src_nullable_column->data_column().get());
+        auto& dest_nullable_column = down_cast<NullableColumn&>(*dest_column);
+        auto& dest_null_data = down_cast<NullableColumn&>(*dest_column).null_column_data();
+        auto& dest_data_column = down_cast<ArrayColumn&>(*dest_nullable_column.data_column());
+
+        dest_null_data = src_nullable_column->immutable_null_column_data();
+        dest_nullable_column.set_has_null(src_nullable_column->has_null());
+
+        if (src_nullable_column->has_null()) {
+            for (size_t i = 0; i < chunk_size; i++) {
+                if (!src_nullable_column->is_null(i)) {
+                    _array_distinct_item<PT, HashSet>(*src_data_column, i, &hash_set, &dest_data_column);
+                    hash_set.clear();
+                } else {
+                    dest_data_column.append_default();
+                }
+            }
+        } else {
+            for (size_t i = 0; i < chunk_size; i++) {
+                _array_distinct_item<PT, HashSet>(*src_data_column, i, &hash_set, &dest_data_column);
+                hash_set.clear();
+            }
+        }
+    } else {
+        const auto* src_data_column = down_cast<const ArrayColumn*>(src_column.get());
+        auto* dest_data_column = down_cast<ArrayColumn*>(dest_column.get());
+
+        for (size_t i = 0; i < chunk_size; i++) {
+            _array_distinct_item<PT, HashSet>(*src_data_column, i, &hash_set, dest_data_column);
+            hash_set.clear();
+        }
+    }
+    return dest_column;
+}
+
+ColumnPtr ArrayFunctions::array_distinct_boolean(FunctionContext* context, const Columns& columns) {
+    return ArrayFunctions::template array_distinct<TYPE_BOOLEAN, phmap::flat_hash_set<int8_t, StdHash<int8_t>>>(
+            columns);
+}
+
+ColumnPtr ArrayFunctions::array_distinct_tinyint(FunctionContext* context, const Columns& columns) {
```

Comment:
it is some tedious,

suggest to use template class instead of template function

template<PrimitiveType PT> 
struct ArrayFunctionStruct {
using CppType = RuntimeCppType<PT>;
public static ColumnPtr array_distinct(FunctionContext* context, const Columns& columns) {
    return ArrayFunctions::template array_distinct<PT, phmap::flat_hash_set<CppType, StdHash<CppType>>>(columns);
}
};

Reply:
Fixed